packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp

```diff
@@ -1587,59 +1802,64 @@ class FastILUPrec
         void compute()
         {
             Kokkos::Timer timer;
-            #ifdef FASTILU_TIMER
-            std::cout << "  >> compute <<" << std::endl;
-            Kokkos::Timer Timer;
-            Kokkos::Timer Timer2;
-            #endif
+            FASTILU_CREATE_TIMER(Timer);
             if ((level > 0) && (guessFlag !=0))
             {
                 initGuessPrec->compute();
+                FASTILU_FENCE_REPORT_TIMER(Timer, ExecSpace(), "  > initGuess");
             }
-            #ifdef FASTILU_TIMER
-            ExecSpace().fence();
-            std::cout << "  > initGuess " << Timer.seconds() << std::endl;
-            Timer.reset();
-            #endif
+
             numericILU();
-            #ifdef FASTILU_TIMER
-            ExecSpace().fence();
-            std::cout << "  > numericILU " << Timer.seconds() << std::endl;
-            Timer.reset();
-            #endif
+            FASTILU_FENCE_REPORT_TIMER(Timer, ExecSpace(), "  > numericILU ");
+
             FastILUFunctor<Ordinal, Scalar, ExecSpace> iluFunctor(aRowMap_[nRows], blkSzILU,
                     aRowMap, aRowIdx, aColIdx, aVal,
-                    lRowMap, lColIdx, lVal, uRowMap, uColIdx, uVal, diagElems, omega);
+                    lRowMap, lColIdx, lVal, uRowMap, uColIdx, uVal, diagElems, omega, blockCrsSize, level);
             Ordinal extent = aRowMap_[nRows]/blkSzILU;
             if (aRowMap_[nRows]%blkSzILU != 0)
             {
                 extent++;
             }
-            //Ordinal extent = aRowMap[nRows];
-            //ExecSpace().fence();
 
-            for (int i = 0; i < nFact; i++) 
+            for (int i = 0; i < nFact; i++)
             {
-                Kokkos::parallel_for(Kokkos::RangePolicy<ExecSpace>(0, extent), iluFunctor);
+#ifdef FASTILU_ONE_TO_ONE_UNBLOCKED
+              // Force serialization to avoid races but still perform operation on device
```

Comment:
I confirmed the race in the original impl. I think this is similar to par_ilut where the race is part of the algorithm.